### PR TITLE
Add Log-in functionality to application

### PR DESCRIPTION
**Pull request by mathewbaltes**

### CODE REVIEW

File: test/controllers/users_controller_test.rb [hunk 1]
```diff
@@ -21,4 +23,38 @@ class UsersControllerTest < ActionDispatch::IntegrationTest
     assert_select 'div.form-control-feedback', count: 1
     assert_select 'input.js-save-user'
   end
+
+  test 'cannot POST a new user when logged in' do
+    assert_difference 'User.count', 1 do
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Consider utilizing a method to simply login rather than have this post as it appears that you're trying to test this part.

File: test/flows/sessions_crud_test.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+require 'flow_test_helper'
+
+class SessionsCrudTest < FlowTestCase
+  test 'login a user' do
+    User.create(name: 'Mathew', email: 'test@test.com', password: 'asdf123', password_confirmation: 'asdf123')
```
Comment (bboe) [inline test/flows/sessions_crud_test.rb]: What happens if this action fails?

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
+class SessionsController < ApplicationController
+  def create
+    user = User.find_by(email: params[:session][:email].downcase)
```
Comment (mathewbaltes) [inline app/controllers/sessions_controller.rb]: No, it does not :)

File: config/routes.rb [hunk 4]
```diff
@@ -5,5 +5,10 @@
     post :share, on: :member
   end
 
+  # routes for login
+  get 'login', to: 'sessions#new'
```
Comment (mathewbaltes) [inline config/routes.rb]: It was a suggestion from the tutorial, however I think that in this circumstance, logout mapped to :destroy is a little more clear as a routing choice.

Comment (mathewbaltes) [inline config/routes.rb]: I'll refactor this since I was able to achieve similar behavior using the collection for 'logout'

File: test/controllers/images_controller_test.rb [hunk 5]
```diff
@@ -28,7 +28,7 @@ def setup
     assert_select 'span.js-tag', 'test'
     assert_select 'span.js-tag', count: 1
     assert_select 'a.js-tag-link', count: 1
-    assert_select "a.share-image[href='#{new_image_path}']"
+    assert_select 'a.js-share-image', count: 2
```
Comment (mathewbaltes) [inline test/controllers/images_controller_test.rb]: This was an unintentional change, fixed.

File: test/controllers/sessions_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
+require 'test_helper'
+
+class SessionsControllerTest < ActionDispatch::IntegrationTest
+  test 'valid user with proper credentials authenticates correctly and can log out' do
```
Comment (mathewbaltes) [inline test/controllers/sessions_controller_test.rb]: Agreed, I have separated these into two tests.

File: test/controllers/users_controller_test.rb [hunk 7]
```diff
@@ -7,6 +7,8 @@ class UsersControllerTest < ActionDispatch::IntegrationTest
         user: { name: 'Mathew', email: 'a@a.com', password: 'blah123', password_confirmation: 'blah123' }
       }
     end
+
+    assert logged_in?
```
Comment (mathewbaltes) [inline test/controllers/users_controller_test.rb]: Agreed, I have updated this to reflect the correct intentions.

File: test/controllers/images_controller_test.rb (hunk 5 above)
Comment (mathewbaltes) [inline test/controllers/images_controller_test.rb]: Refactored to test visibility of buttons on the image controller. Should I split this into a separate commit.

Comment (bboe) [inline test/controllers/images_controller_test.rb]: I simply wasn't sure why it changed from looking for a specific url to looking for two such links. Do you think the presence of the links on the page is a sufficient assertion?

File: app/helpers/nav_bar_link_helper.rb [hunk 8]
```diff
@@ -1,8 +1,8 @@
 module NavBarLinkHelper
-  def nav_bar_link(text, path, extra_anchor_classes = [])
+  def nav_bar_link(text, path, extra_anchor_classes = [], method = nil)
```
Comment (bboe) [inline app/helpers/nav_bar_link_helper.rb]: Great update on making this its own commit.

File: app/views/users/login_show.html.erb [hunk 9]
```diff
@@ -0,0 +1,13 @@
+<%= simple_form_for @user, url: users_path do |f| %>
```
Comment (bboe) [inline app/views/users/login_show.html.erb]: What is this file for?

File: test/controllers/sessions_controller_test.rb [hunk 10]
```diff
@@ -0,0 +1,50 @@
+require 'test_helper'
+
+class SessionsControllerTest < ActionDispatch::IntegrationTest
+  test 'valid user with proper credentials authenticates correctly' do
+    User.create!(name: 'test', email: 'test@test.com', password: 'asdf123', password_confirmation: 'asdf123')
```
Comment (bboe) [inline test/controllers/sessions_controller_test.rb]: Consider using a factory here that provides default valid arguments so that if you ever update the model to have more required parameters you do not have to update all the test cases that utilize the model.

File: test/controllers/sessions_controller_test.rb [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    assert_equal 'Welcome test!', flash[:success]
+    assert_redirected_to root_path
+  end
+
+  test 'properly authenticated user can log out' do
+    User.create!(name: 'test', email: 'test@test.com', password: 'asdf123', password_confirmation: 'asdf123')
+
+    post sessions_path, params: {
```
Comment (bboe) [inline test/controllers/sessions_controller_test.rb]: Consider making a private function for "login_as" or something to that effect so it's clear that's what this part of the test code is doing. Also it's good to assert that the login works as expected (assert_response).

File: test/flows/sessions_crud_test.rb [hunk 12]
```diff
@@ -0,0 +1,19 @@
+require 'flow_test_helper'
```
Comment (bboe) [inline test/flows/sessions_crud_test.rb]: Is this a flow worth testing in isolation? In other words, what's the benefit of this flow over the controller test?

File: test/controllers/images_controller_test.rb (hunk 5 above)
Comment (mathewbaltes) [inline test/controllers/images_controller_test.rb]: Agreed, made them more specific.

File: app/views/users/login_show.html.erb (hunk 9 above)
Comment (mathewbaltes) [inline app/views/users/login_show.html.erb]: Extra file, nice catch.

File: test/flows/sessions_crud_test.rb (hunk 12 above)
Comment (mathewbaltes) [inline test/flows/sessions_crud_test.rb]: No, it isn't. Probably duplicated functionality. -removing

File: test/factories/user_factory.rb [hunk 13]
```diff
@@ -0,0 +1,7 @@
+module TestFactory
```
Comment (bboe) [inline test/factories/user_factory.rb]: This is even more awesome than what I was thinking.
